---
title: "expression analyses with eQTL data"
output: html_notebook
---
Goal: 
use previously generated table of gametologs, repeat male v. female expression analyses and mat and pat allele expression analyses (ase) using the large eQTL population dataset

#### Packages & such
```{r packages, echo=TRUE, message=FALSE, warning=FALSE}
library(tidyverse)
library(DESeq2)
library(gtools)
`%nin%` <- negate(`%in%`)
get_obs_table <- 
  function(data, cutoff) {
  data %>%
    filter(res.allele.padj < 0.1, abs(res.allele.log2FoldChange) > cutoff) %>%
    summarise(n_Y_oe = sum(res.allele.log2FoldChange > 0),
              n_X_oe = sum(res.allele.log2FoldChange <= 0)) %>%
    mutate(foldChangeCutoff = cutoff)
  }
```

#### Raw read count data
```{r load-readcounts, message=FALSE, warning=FALSE, include=FALSE}
rnaCounts<-read_table("readCountsRNAeQTL/rhastTXeQTLrna.txt")
dnaCounts<-read_table("readCountsDNAeQTL/rhastTXeQTLdna.txt")
```

#### data wrangling
- fix sample names
- order by sex
- compile sample information for deseq2
- remove sample 40aFLD and 24fMLD - first one is abberrant, second has weirdly small fastq

```{r}
rnaColnames<-colnames(rnaCounts)
rnaColnamesRepaired <- gsub(".*NEBNext_dual_i._.+\\.([0-9]+[a-z]+[FM])LRAligned.+", "\\1", rnaColnames)
colnames(rnaCounts)<-rnaColnamesRepaired 


dnaColnames<-colnames(dnaCounts) 
dnaColnamesRepaired<- gsub("NS\\.[0-9]+\\.[0-9]+.IDT_i._\\d+---IDT_i._\\d+\\.(\\d+[a-z]+[FM])LD_sortMarkDups.*", "\\1", dnaColnames)
colnames(dnaCounts)<-dnaColnamesRepaired

setdiff(rnaColnamesRepaired,dnaColnamesRepaired)
# remove samples "67eF" "14bF" "45aF"
# also need to remove 40aF and 24fM 
# 40aF< and 35aM have aberrant coverage patterns
```


```{r}

```

```{r}
clean_counts <- function(data, columns_to_remove) {
  # Calculate number of columns that contain "F" and "M"
  num_female_cols <- ncol(select(data, contains("F", ignore.case = FALSE)))
  num_male_cols <- ncol(select(data, contains("M", ignore.case = FALSE)))
  num_cols<-num_female_cols+num_male_cols

  data %>%
    select(-all_of(columns_to_remove)) %>%  # Remove specified columns
    relocate(1:6, ends_with("F"), ends_with("M")) %>%  # Sort columns
    mutate(Chr = case_when(  # Relabel chromosomes
      grepl("X", Chr) ~ "X",
      grepl("Y", Chr) ~ "Y",
      grepl("A1", Chr) ~ "A1",
      grepl("A2", Chr) ~ "A2",
      grepl("A3", Chr) ~ "A3",
      grepl("A4", Chr) ~ "A4",
      TRUE ~ NA_character_
    )) %>%
    rowwise() %>%
    mutate( # Calculate mean counts
      femaleCountsTotal = sum(c_across(contains("F", ignore.case = FALSE))),  # Calculate total female counts
      female_mean = (femaleCountsTotal / num_female_cols),  # Calculate mean female counts dynamically
      maleCountsTotal = sum(c_across(contains("M", ignore.case = FALSE))),  # Calculate total male counts
      male_mean = (maleCountsTotal / num_male_cols),  # Calculate mean male counts dynamically
    totalCountsGene = (femaleCountsTotal+maleCountsTotal),
    meanCounts = totalCountsGene / num_cols) %>%
    relocate(1:6, totalCountsGene, meanCounts, femaleCountsTotal, female_mean, maleCountsTotal, male_mean)  # Reorder columns
}
#test_coutns<-clean_counts(rnaCounts,rnacolsremove)
```

```{r}
rnacolsremove<-c("67eF","14bF","45aF","40aF","24fM","35aM")
dnacolsremove<-c("40aF","24fM","35aM")
rnaCountsClean <- clean_counts(rnaCounts,rnacolsremove) # get total gene-level counts for filtering
dnaCountsClean <- clean_counts(dnaCounts,dnacolsremove)
```


```{r}
# 156 samples
setdiff(colnames(rnaCountsClean),colnames(dnaCountsClean))
# ok all g
#rnaCountsWide<-rnaCountsClean %>% pivot_longer(names_to = )

dnaSamplenamesClean<-colnames(select(dnaCountsClean,-1:-12))
sampleInfo<-as_tibble_col(dnaSamplenamesClean, column_name="Sample") %>% mutate(.,Sex = if_else(grepl("*M", Sample),"M", "F"))

```

### DNA read ratio estimation
do this first. For filtering reasons! We want to remove sites with high mapping bias
This will also help eliminate genes where there's weird amounts of mapping on Y for females. Hopefully!

### load X-Y gametologs
```{r include=FALSE}
pgMatOrths<-read_csv("data/pgMatOrths.csv")
```
### DNA read ratios

```{r}
### filter genes with sig DNA male coverage bw x and y gametologs instead of just chucking out "badgenes"
eqtl_x_dna<- filter(dnaCountsClean, Chr =="X") %>% select(c(Geneid,totalCountsGene:161)) ## extract X gene counts
eqtl_y_dna<- filter(dnaCountsClean, Chr =="Y") %>% select(c(Geneid,totalCountsGene:161)) ## extract Y gene counts

reads_pg_dna<-left_join(pgMatOrths,eqtl_x_dna, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y_dna, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>% 
  drop_na() %>% select(-flag_tx_pat,-flag_tx_mat) %>% 
  unique() %>%
  rowwise() %>% 
  mutate(totalCountAll=(totalCountsGene.mat+totalCountsGene.pat)) %>%
  ungroup() %>% filter(totalCountAll >20) %>%
  relocate(c(pairID:male_mean.mat, totalCountsGene.pat:male_mean.pat,totalCountAll)) 


colnames(reads_pg_dna)

reads_pg_dna_noPAR<-reads_pg_dna %>% filter(start_tx_pat > 45000000)

ggplot(reads_pg_dna,aes(x=male_mean.mat,y=male_mean.pat)) + geom_point() +
    geom_abline(slope = 1, intercept = 0, linetype=1) +
  theme_bw()+
  geom_smooth(method ="lm") +
  xlab("mean dna read counts of X gametolog") +
  ylab("mean dna read counts of Y gametolog") 
ggplot(reads_pg_dna_noPAR,aes(x=male_mean.mat,y=male_mean.pat)) + geom_point() +
    geom_abline(slope = 1, intercept = 0, linetype=1) +
  theme_bw()+
  geom_smooth(method ="lm") +
  xlab("mean dna read counts of X gametolog") +
  ylab("mean dna read counts of Y gametolog") 
## same basically. but good to remove.

```
### explore DNA output
```{r}
ggplot(dnaCountsClean, aes(meanCounts))+geom_histogram() + facet_wrap(~ Chr)
ggplot(dnaCountsClean, aes(female_mean))+geom_histogram() + facet_wrap(~ Chr)
ggsave("femaleMeanCounts_nolim_hist.pdf")
ggplot(dnaCountsClean, aes(female_mean))+geom_histogram() + xlim(0,200)+ facet_wrap(~ Chr)
ggsave("femaleMeanCounts_lim200_hist.pdf")
ggplot(dnaCountsClean, aes(male_mean))+geom_histogram() + xlim(0,200)+ facet_wrap(~ Chr)
ggsave("maleMeanCounts_lim200_hist.pdf")
ggplot(dnaCountsClean, aes(male_mean))+geom_histogram() + facet_wrap(~ Chr)
ggsave("maleMeanCounts_nolim_hist.pdf")
```

```{r}
ase_data_dna<-reads_pg_dna_noPAR %>% select(c(pairID,contains("M", ignore.case = FALSE ))) %>% column_to_rownames(var ="pairID")
## ase matrix
IDs<-colnames(ase_data_dna)
sorted_col_names <- mixedsort(IDs)
ase_data_dna <- ase_data_dna[, sorted_col_names]
#colnames(ase_data)
# yay!!
sample<-str_remove(sorted_col_names,"..at") # tm1 tm1 etc.
allele<-str_extract(sorted_col_names,".at$") # mat pat mat pat

colInfo<-data.frame(sorted_col_names,sample,allele) # wooo

ase_matrix_dna<-as.matrix(ase_data_dna)
design<- ~sample+allele

m <- 75

dds <- DESeqDataSetFromMatrix(ase_data_dna, colInfo, design)
sizeFactors(dds) <- rep(1, 2*m)
dds <- DESeq(dds, fitType="local")

resultsNames(dds)

res.allele <- results(dds, name="allele_pat_vs_mat")
head(res.allele$log2FoldChange)

adj_pvals_dna<-res.allele$padj


pvals_genes_dna<-data.frame(res.allele@rownames,res.allele$padj,res.allele$log2FoldChange,res.allele$lfcSE) #%>%
reads_pg_pvals_dna<-inner_join(reads_pg_dna_noPAR, pvals_genes_dna, by =c("pairID" ="res.allele.rownames")) %>%
  mutate(sigTest = 
           case_when((res.allele.padj < 0.1) ~ "sig",
                     TRUE ~"nonsig" )) %>% relocate(c(pairID:totalCountAll,res.allele.padj:sigTest))
write_csv(reads_pg_pvals_dna, "dna_ase_results_eqtl_noPAR.csv")
#head(reads_pg_pvals)

biasedgenes_dna<-filter(reads_pg_pvals_dna, sigTest == "sig" & abs(res.allele.log2FoldChange) >0.5) #%>% select(pairID, id_tx_mat, id_tx_pat)


fold_change_cutoffs <- c(0.5, 1, 1.5, 2)
obs_tables_dna <- lapply(fold_change_cutoffs, function(cutoff) {
  get_obs_table(reads_pg_pvals_dna, cutoff)
})
combined_table_dna <- do.call(rbind, obs_tables_dna)
print(combined_table_dna)
write_csv(combined_table_dna, "n_xy_oe_DNA_eQTL_lfcRanges.csv")

badgenes_dna<-reads_pg_dna_noPAR %>% 
  filter(female_mean.pat >= 20 & chr_tx_pat =="Y") %>% select(pairID,id_tx_mat,id_tx_pat)

```


#### Expression analyses
RNA read counts.
```{r}
badgenes<-rnaCountsClean %>% 
  filter(female_mean>= 20 & Chr =="Y") %>% select(Geneid) # get genes where females have too much Y coverage
# some may be on Y PAR - which is fromm 0-45MB. Disregard and remove all for now!
badgenes_x <- pgMatOrths %>% select(id_tx_mat,id_tx_pat) %>% filter(id_tx_pat%in%badgenes$Geneid) %>% select(id_tx_mat)
## find corresponding X gene for every "bad" Y gene
shared_bad_biased<-intersect(badgenes$Geneid,biasedgenes_dna$id_tx_pat) # 327 genes
shared_bad<-intersect(badgenes$Geneid,badgenes_dna$id_tx_pat) # 400 genes
```


```{r}
readsMatAll<-rnaCountsClean %>% column_to_rownames("Geneid") %>% select("39dF":160) # convert to matrix of all genes and inds

readsMatAuto<-rnaCountsClean %>% filter(grepl("A",Chr)) %>% # matrix with just autosomes, for generating norm factors
   column_to_rownames("Geneid") %>% select("39dF":160) 

sampleNames<- colnames(readsMatAll)
#
sampleInfo<-data.frame(sampleNames) %>% 
  mutate(.,Sex = if_else(grepl("*M", sampleNames),"M", "F"))

df_all<-DESeqDataSetFromMatrix(readsMatAll,sampleInfo,design = ~ 1)
df_auto<-DESeqDataSetFromMatrix(readsMatAuto,sampleInfo,design = ~ 1)
dds_all <- DESeq(df_all)
dds_auto <- DESeq(df_auto)
allchrFactors<-sizeFactors(dds_all)
autoFactors<-sizeFactors(dds_auto)
compareFactors<-data.frame(allchrFactors,autoFactors) %>% rownames_to_column("sample") 
compareFactorsLong <- compareFactors %>% pivot_longer(cols = -sample, names_to = "normType", values_to = "value")
ggplot(compareFactorsLong) + geom_point(aes(x = sample, y = value, color = normType)) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

#ggsave("compare_norm_factors.pdf", h = 5, w = 18)
## very similar
## use autosome factors anyway *for dc results


```

```{r}
# load set of hemizygous genes
tx_yloss_nc_overlap<-read.csv("data/tx_yloss_overlapNCcomplete_loss.csv")


# replace og coutn tabels size factors
dds_all$sizeFactor<-autoFactors
dds_all$sizeFactor

normalized_counts<-counts(dds_all, normalized=TRUE)
normcounts<-as.data.frame(normalized_counts) %>% rownames_to_column()
#write_csv(normcounts,"norm_counts_autosomefactors_eQTLleafRhast.csv")
normcounts<-read_csv("data/norm_counts_autosomefactors_eQTLleafRhast.csv.gz")
hemicounts <- as.data.frame(normalized_counts) %>% 
  rownames_to_column(var = "Geneid") %>% filter(Geneid%in%tx_yloss_nc_overlap$repGene) %>%
  rowwise %>% 
  mutate(maleMean = mean(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(maleVar = sd(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(femaleMean = mean(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(femaleVar = sd(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(totalCount =sum(c_across("39dF":"43bM"))) %>%
  mutate(totalVar = sd(c_across("39dF":"43bM"))) %>%
  ungroup() %>%
  filter(totalCount >20)


ggplot(hemicounts, aes(x = femaleMean, y = maleMean)) + 
  geom_point(aes(size = femaleVar), alpha = 0.5) +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  scale_x_continuous(expand = c(0, 0), limits = c(0, 20000),breaks =seq(0,18000,2000)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, 20000),breaks =seq(0,18000,2000)) +
  xlab("Average female expression") +
  ylab("Average male expression")


ggplot(hemicounts, aes(x = femaleMean, y = maleMean)) + 
  geom_point(aes(size = maleVar), alpha = 0.5) +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  scale_x_continuous(expand = c(0, 0), limits = c(0, 20000),breaks =seq(0,18000,2000)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, 20000),breaks =seq(0,18000,2000)) +
  xlab("Average female expression") +
  ylab("Average male expression")

ggplot(hemicounts, aes(x = femaleMean, y = maleMean)) + 
  geom_point(aes(size = totalVar), alpha = 0.5) +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  scale_x_continuous(expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,18000,2000)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,18000,2000)) +
  xlab("Average female expression") +
  ylab("Average male expression")

ggplot(hemicounts, aes(x = femaleMean, y = maleMean)) + 
  geom_point() +
  geom_errorbar(aes(ymin = maleMean-maleVar, ymax = maleMean+maleVar)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleVar, xmax = femaleMean+femaleVar)) +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  #scale_x_continuous(expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) + 
  #scale_y_continuous(expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) +
  xlab("Average female expression") +
  ylab("Average male expression")

ggsave("hemizygous_eqtldata_dosage_errorbars.jpg", h = 5, w = 7)
```



## ASE


```{r}
eqtl_x<- filter(rnaCountsClean, Chr =="X") %>% select(c(Geneid,totalCountsGene:"43bM")) ## extract X gene counts
eqtl_y<- filter(rnaCountsClean, Chr =="Y") %>% select(c(Geneid,totalCountsGene:"43bM")) ## extract Y gene counts

reads_pg_nofilter<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>%
  drop_na()%>%  summarise(n = n())

reads_pg_PARfilter<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>%
  filter(start_tx_pat > 45000000) %>% summarise(n = n())

reads_pg_badRNAgenesFilter<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>%
  filter(id_tx_mat%nin%badgenes_x$id_tx_mat) %>% # weird expression filter
  filter(id_tx_pat%nin%badgenes$Geneid) %>%
  filter(start_tx_pat > 45000000) %>%
  summarise(n = n())

reads_pg_badDNAgenesFilter<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>%
  filter(id_tx_mat%nin%badgenes_x$id_tx_mat) %>% # weird expression filter
  filter(id_tx_pat%nin%badgenes$Geneid) %>%
  filter(start_tx_pat > 45000000) %>%
  #filter(pairID%nin%biasedgenes_dna$pairID) %>% # biased gene filter
  filter(pairID%nin%badgenes_dna$pairID) %>%
  summarise(n = n())

reads_pg_biasedDNAgenesFilter<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>%
  filter(id_tx_mat%nin%badgenes_x$id_tx_mat) %>% # weird expression filter
  filter(id_tx_pat%nin%badgenes$Geneid) %>%
  filter(start_tx_pat > 45000000) %>%
  filter(pairID%nin%biasedgenes_dna$pairID) %>% # biased gene filter
  filter(pairID%nin%badgenes_dna$pairID) %>%
  summarise(n = n())
 
reads_pg<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>% 
  filter(id_tx_mat%nin%badgenes_x$id_tx_mat) %>% # weird expression filter
  filter(id_tx_pat%nin%badgenes$Geneid) %>%
  filter(start_tx_pat > 45000000) %>% # par filter
  filter(pairID%nin%biasedgenes_dna$pairID) %>% # biased gene filter
  filter(pairID%nin%badgenes_dna$pairID) %>% # filter weird Y mapping
  drop_na() %>% select(-flag_tx_pat,-flag_tx_mat) %>% 
  unique() %>%
  rowwise() %>% 
  mutate(totalCountAll=(totalCountsGene.mat+totalCountsGene.pat)) %>%
  ungroup() %>% filter(totalCountAll >20)

reads_pg_allfilters<-reads_pg %>% summarise(n=n())
# 267 genes remain :()
```
  
Attempting fishers exact test instead...
we'll see
```{r}
# general idea
# 
#     X  Y
# RNA counts 
# DNA

fisher <- function(a,b,c,d){
  data <- matrix(c(a,b,c,d),ncol=2)
  c(p = fisher.test(data)$p.value) 
}

```
  

```{r}
# table of read numbers
category<-c("reads_pg_nofilter","reads_pg_PARfilter","reads_pg_badRNAgenesFilter","reads_pg_badDNAgenesFilter","reads_pg_biasedDNAgenesFilter","reads_pg_allfilters")
counts<-c(reads_pg_nofilter$n,reads_pg_PARfilter$n,reads_pg_badRNAgenesFilter$n,reads_pg_badDNAgenesFilter$n,reads_pg_biasedDNAgenesFilter$n,reads_pg_allfilters$n)

filtering_summary<-data.frame(category,counts)
print(filtering_summary)

ggplot(reads_pg,aes(x=male_mean.mat,y=male_mean.pat)) + geom_point() +
    geom_abline(slope = 1, intercept = 0, linetype=1) +
  theme_bw()+
  geom_smooth(method ="lm") +
  xlab("mean exp of X gametolog") +
  ylab("mean exp of Y gametolog") +
  scale_x_continuous(transform = "log2") +
  scale_y_continuous(transform = "log2")
ggsave("gametolog_readCounts_all_may27_logscale.jpg", h = 5, w = 7)

```
deseq ase test
```{r}
ase_data<-reads_pg %>% select(c(pairID,contains("M", ignore.case = FALSE ))) %>% column_to_rownames(var ="pairID")
## ase matrix
IDs<-colnames(ase_data)
library(gtools)
sorted_col_names <- mixedsort(IDs)
ase_data <- ase_data[, sorted_col_names]
#colnames(ase_data)
# yay!!
sample<-str_remove(sorted_col_names,"..at") # tm1 tm1 etc.
allele<-str_extract(sorted_col_names,".at$") # mat pat mat pat

colInfo<-data.frame(sorted_col_names,sample,allele) # wooo

ase_matrix<-as.matrix(ase_data)
design<- ~sample+allele

m <- 75

dds <- DESeqDataSetFromMatrix(ase_data, colInfo, design)
sizeFactors(dds) <- rep(1, 2*m)
dds <- DESeq(dds, fitType="local")

resultsNames(dds)

res.allele <- results(dds, name="allele_pat_vs_mat")
head(res.allele$log2FoldChange)

adj_pvals<-res.allele$padj


pvals_genes<-data.frame(res.allele@rownames,res.allele$padj,res.allele$log2FoldChange,res.allele$lfcSE) #%>%
reads_pg_pvals<-inner_join(reads_pg, pvals_genes, by =c("pairID" ="res.allele.rownames"))%>%
  mutate(sigTest = 
           case_when(res.allele.padj < 0.1 & abs(res.allele.log2FoldChange) > 1 ~ "sig",
                     TRUE ~"nonsig" )) %>% 
  relocate(c(pairID:male_mean.mat, meanCounts.pat:male_mean.pat, res.allele.padj:sigTest))

#write_csv(reads_pg_pvals, "rna_ase_results_eqtl_may17.csv")
reads_pg_pvals<-read_csv("data/rna_ase_results_eqtl_may17.csv")
head(reads_pg_pvals)
```


```{r}
ggplot(reads_pg_pvals) +
  geom_point(aes(x=start_tx_pat, 
                 y = 2^res.allele.log2FoldChange/(1+2^res.allele.log2FoldChange), size = res.allele.lfcSE,
                 color = sigTest),alpha = 0.5) 
#ggsave("lfc_ase_log2_1.pdf")
ggplot(reads_pg_pvals) +
  geom_point(aes(y=(maleCountsTotal.pat+maleCountsTotal.mat)/(femaleCountsTotal.mat+femaleCountsTotal.pat), 
                 x = maleCountsTotal.pat/(maleCountsTotal.mat),
                 color = sigTest)) +ylim(0,1) + xlim(0,1) +geom_abline(intercept = 0.5, slope = 0.5)
#ggsave("DC_expected_counts_Lfc1_axis.pdf")

# fraction of genes ow or ue on Y depending on stringency
# more stringent lgo2fc2 change, more reduced Y exp, but fewer sig sites
```
### read ratio, not using fold change ratio

```{r}
#df<- reads_pg_pvals %>% mutate(readratio = (male_mean.pat/male_mean.mat))

#ggplot(df)+ geom_histogram(aes(x = (readratio), color = sigTest)) #+xlim(0,1)
reads_pg_pvals2 <- reads_pg_pvals %>%   mutate(sigLFC = 
           case_when(res.allele.padj < 0.1 & abs(res.allele.log2FoldChange) > 1 ~ "True",
                     TRUE ~"False" ))
reads_pg_pvals2$sigLFC<-factor(reads_pg_pvals2$sigLFC, levels = c("True","False"))
mycolors<-RColorBrewer::brewer.pal(2,"Set1")[2:3]

ggplot(reads_pg_pvals2,aes(x=male_mean.mat,y=male_mean.pat, color = sigLFC)) + geom_point(aes(size = res.allele.lfcSE),alpha =0.60) +
    geom_abline(slope = 1, intercept = 0, linetype=1) +
  theme_bw()+
  geom_smooth(method ="lm") +
  xlab("mean expression of X\n gametolog in males") +
  ylab("mean expression of Y\n gametolog in males") +
  scale_x_continuous(transform = "log2",limits = c(NA,4000),breaks = c(0.0625,2,64,2048)) +
  scale_y_continuous(transform = "log2",limits = c(NA,4000),breaks = c(0.0625,2,64,2048)) +
  labs(color= "p-val < 0.1 & |log2FC| >1 ", size = "log2FC std. error") +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        legend.text = element_text(size =12),
        legend.title = element_text(size =12)) 
  
 #scale_color_manual(values = mycolors)
  #ggsave("gametolog_expression_lfc1_fdr0.1.pdf")
#ggsave("gametolog_expression_lfc1_fdr0.1_lfcSE.pdf")
ggsave("plots/gametolog_expression_lfc1_fdr0.1_lfcSE_logscale.pdf",h =6, w=10)


```

```{r}
ggplot(filter(reads_pg_pvals, 
              sigTest == "sig", res.allele.log2FoldChange < 0),
       aes(x=female_mean.mat, y=male_mean.mat)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_smooth(method ="lm") +
   #scale_x_continuous(expand = c(0, 0), limits = c(0, NA) ,breaks =seq(0,16000,2000)) + 
  #scale_y_continuous(expand = c(0, 0), limits = c(0, NA),breaks =seq(0,16000,2000)) +
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic()
### barely any sig ase??

ggplot(filter(reads_pg_pvals, res.allele.log2FoldChange > 0),
       aes(x=female_mean.mat, y=male_mean.mat)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_smooth(method ="lm") +
   #scale_x_continuous(expand = c(0, 0), limits = c(0, NA) ,breaks =seq(0,16000,2000)) + 
  #scale_y_continuous(expand = c(0, 0), limits = c(0, NA),breaks =seq(0,16000,2000)) +
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic()
### add error bar,
# stic
```
nearly done!
- add DNA
- sig testing?
- outgroup
- Ks vs y/x
- blast tx missing genes to tx genome????
  - NC shared set is likely conservative and good to go!
  - haven't id'd partial loss..
- non-one-one orthologs - explore 
- characterise these other genes
- GO GO enrichment

Need table
- how many Y oe vs. X oe genes for each fold change from 0.5 -> 1 -> 2 ... 
- at alpha of FDR 0.1
```{r}



fold_change_cutoffs <- c(0.5, 1, 1.5, 2)
obs_tables <- lapply(fold_change_cutoffs, function(cutoff) {
  get_obs_table((filter(reads_pg_pvals, res.allele.padj < 0.1)), cutoff)
})

combined_table <- do.call(rbind, obs_tables)
print(combined_table)
write_csv(combined_table, "n_xy_oe_RNA_eQTL_lfcRanges.csv")


```
## dosage compensation in "silenced" genes

```{r}
### silenced-ish genes
silenced_set<-filter(reads_pg_pvals, res.allele.padj < 0.1 & res.allele.log2FoldChange < 1) %>% select(id_tx_mat)

silenced_counts <- as.data.frame(normalized_counts) %>% 
  rownames_to_column(var = "Geneid") %>% filter(Geneid%in%silenced_set$id_tx_mat) %>%
  rowwise %>% 
  mutate(maleMean = mean(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(maleVar = sd(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(femaleMean = mean(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(femaleVar = sd(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(totalCount =sum(c_across("39dF":"43bM"))) %>%
  mutate(totalVar = sd(c_across("39dF":"43bM"))) %>%
  ungroup() %>%
  filter(totalCount >20)

 
ggplot(silenced_counts, aes(x = femaleMean, y = maleMean)) + 
  geom_point() +
  geom_errorbar(aes(ymin = maleMean-maleVar, ymax = maleMean+maleVar)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleVar, xmax = femaleMean+femaleVar)) +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  #scale_x_continuous(expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) + 
  #scale_y_continuous(expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) +
  xlab("Average female expression") +
  ylab("Average male expression")

ggsave("silenced_yunderexpressed_dosagecomp.jpg",w = 7, h = 5)
```

Dosage compensation plot combined
```{r}
#completely_silenced<-filter(reads_pg_pvals, res.allele.padj < 0.1 & res.allele.log2FoldChange < 1 & male_mean.pat == 0) %>%
#  select(id_tx_mat)
x_overexpressed<-filter(reads_pg_pvals, res.allele.padj < 0.1 & res.allele.log2FoldChange < 1) %>%
  select(id_tx_mat, id_tx_pat)
y_overexpressed<-filter(reads_pg_pvals, res.allele.padj < 0.1 & res.allele.log2FoldChange > 1) %>% select(id_tx_mat,id_tx_pat)


normcounts_clean<-#as.data.frame(normalized_counts) %>% 
  normcounts %>%
  dplyr::rename("Geneid" ="rowname") %>%
  rowwise %>% 
  mutate(maleMean = mean(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(maleVar = sd(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(femaleMean = mean(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(femaleVar = sd(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(totalCount =sum(c_across("39dF":"43bM"))) %>%
  mutate(totalVar = sd(c_across("39dF":"43bM"))) %>%
  ungroup() %>%
  filter(totalCount >20) #%>%
dc_combo <- normcounts_clean %>% mutate(genetype = case_when(
   #Geneid%in%completely_silenced$id_tx_mat ~ "completeYsilenced",
    Geneid%in%x_overexpressed$id_tx_mat ~ "xOverexp",
    #Geneid%in%y_overexpressed$id_tx_mat ~ "yOverexp",
    Geneid%in%tx_yloss_nc_overlap$repGene ~ "hemizygous", TRUE ~ "other")) %>% 
    filter(genetype != "other")


ggplot(dc_combo, aes(x = femaleMean, y = maleMean, color = genetype)) + 
  geom_point(alpha = 0.5, size = 3) +
  geom_errorbar(aes(ymin = maleMean-maleVar, ymax = maleMean+maleVar, color = genetype)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleVar, xmax = femaleMean+femaleVar, color = genetype)) +
  geom_smooth(method="lm", fullrange = TRUE) +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +
  theme_bw() +
  xlab("Average female expression") +
  ylab("Average male expression") +
  labs(color= NULL) +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        legend.text= element_text(size =12)) +
  scale_color_discrete(labels = c("hemizygous"="Hemizygous", "xOverexp" = "Partial or complete\n Y silencing"))
ggsave("plots/combined_dc_plot_hemi_ysilence.jpg", h = 6, w =9)
```


Messing around w KS plots
```{r}
ks_raw<-read_table("Rhastxy.april2024.stephen.coge.aligncoords.gcoords.ks.txt", skip = 2) 
ks_colnames<-colnames(ks_raw)
ks_raw<-read_table("Rhastxy.april2024.stephen.coge.aligncoords.gcoords.ks.txt", comment = "#", col_names = ks_colnames)
mydelim = "\\|\\|"
col_parts_1 <- as.vector(unlist(str_split(colnames(ks_raw)[4],  mydelim)))
col_parts_2<- as.vector(unlist(str_split(colnames(ks_raw)[8],  mydelim)))


# Apply the function to the dataframe
ks_data_separated <- ks_raw %>% separate(
  col = `chr1||start1||stop1||name1||strand1||type1||db_feature_id1||genome_order1||percent_id1`, 
                                         sep = mydelim, into = col_parts_1) %>%
  separate(col = `chr2||start2||stop2||name2||strand2||type2||db_feature_id2||genome_order2||percent_id2`,
           sep = mydelim, into = col_parts_2) %>% select(-"b<db_genome_id>_<chr>",-"a<db_genome_id>_<chr>",-GEVO_link) %>%
  dplyr::rename("Ks"="#Ks") %>% mutate(id_tx_pat = gsub("-RA","",name1)) %>% mutate(id_tx_mat = gsub("-RA","",name2))
# View the modified dataframe
colnames(ks_data_separated)

### combine with y x exp data

reads_pg_pvals_ks <-left_join(reads_pg_pvals, ks_data_separated, by = c("id_tx_mat","id_tx_pat")) %>% 
  relocate(Ks:block_score) %>%
  drop_na() filter

ggplot(reads_pg_pvals_ks) + 
         geom_point(aes(y = Ks, x = male_mean.pat/male_mean.mat)) +
         theme_classic() + xlim(0,2) +ylim(0,0.3) 
ggsave("Ks_readRatio_filter.pdf")
ggsave("Ks_readTa")
ggplot(reads_pg_pvals_ks) + geom_point(aes(y = Ks, x = res.allele.log2FoldChange)) #+ ylim(0,0.5)  +xlim(-10,10)

```
- sliding windows?
- color code by site type (silenced etc.)
```{r}
ks_data_types<- ks_data_separated %>% mutate_at(c('start1', 'start2'), as.numeric) %>%
     mutate(genetype = case_when(
    (id_tx_mat%in%x_overexpressed$id_tx_mat & id_tx_pat%in%x_overexpressed$id_tx_pat) ~ "xOverexp",
    (id_tx_mat%in%y_overexpressed$id_tx_mat & id_tx_pat%in%y_overexpressed$id_tx_pat) ~ "yOverexp", TRUE ~ "other")) %>%  
    filter(Ks < 0.3) %>% filter(chr1  =="Y" & chr2 =="X")

ggplot(ks_data_types) + geom_point(aes(y = Ks, x = start2, color = genetype)) + 
  scale_color_manual(values=alpha(c("grey", "blue", "green"),c(0.2, 0.5, 0.5))) + ylab("Ks between X and Y genes") + xlab("gene start position on X chromosome") +
  theme_classic() + geom_vline(xintercept = 220e6, linetype = "dashed")# scale_x_continuous(breaks = seq(0,500e6,100e6)) +theme_classic()
ggsave("Ks_labels_xcoord.jpg",h=5,w=7)

ggplot(ks_data_types) + geom_point(aes(y = Ks, x = start1, color = genetype)) + 
  scale_color_manual(values=alpha(c("grey", "blue", "green"),c(0.2, 0.5, 0.5))) + ylab("Ks between X and Y genes") + xlab("gene start position on Y chromosome") +
  theme_classic() +
  geom_vline(xintercept = 45e6, linetype = "dashed")
ggsave("Ks_labels_ycoord.pdf",h=5,w=7)

```

genome Ks plot for the hell of it
```{r}
ks_mb<-ks_data_separated %>% mutate_at(c('start1', 'start2'), as.numeric) %>% 
  mutate(start1_mb = (start1/1e6)) %>%
  mutate(start2_mb = (start2/1e6))
library(slider)
windowmaker_1<- function(df,win,ks){df %>% filter(Ks<ks) %>%
  group_by(chr1) %>%
  arrange(name1,chr1, start1_mb) %>%
  mutate(Ks_win = slider::slide_dbl(Ks, median, .before = win/2, .after = win/2,.step = 1,.complete =T))}

windowmaker_2<- function(df,win,ks){df %>% filter(Ks<ks) %>%
  group_by(chr2) %>%
  arrange(name2,chr2, start2_mb) %>%
  mutate(Ks_win = slider::slide_dbl(Ks, median, .before = win/2, .after = win/2,.step = 1,.complete =T))}

ks_win100<-windowmaker_1(ks_mb,win = 100,0.3)
ks_win100_x<-windowmaker_2(ks_mb, win = 100, 0.3)

ks_win50<-windowmaker_1(ks_mb,win = 50,0.3)
ks_win20_x<-windowmaker_2(ks_mb, win = 20, 0.3)
library(RColorBrewer)
ggplot(filter(ks_win50,!grepl("scaffold", chr1)), aes(start1_mb,group = chr1)) +  
   geom_line(aes(y=Ks_win, colour = chr1)) +
   facet_grid(~chr1,scales="free_x") +
   ylim(0,0.05) +
  scale_color_brewer(palette = "Dark2") +
  xlab("Y-bearing Haplotype\nposition(Mb)") +
  ylab("Median Ks\n(100 gene windows)") 

ggsave("Ks_coords_allchr_TX_hap2_20win.jpg", h = 5, w = 9)


ggplot(filter(ks_win20_x,!grepl("scaffold", chr2)), aes(start2_mb,group = chr2)) +  
   geom_line(aes(y=Ks_win, colour = chr2)) +
   facet_grid(~chr2,scales="free_x") +
   ylim(0,0.05) +
  scale_color_brewer(palette = "Dark2") +
  xlab("X-bearing Haplotype\nposition(Mb)") +
  ylab("Median Ks\n(100 gene windows)") 
ggsave("Ks_coords_allchr_TX_hap1_20win.pdf", h = 5, w = 9)

```

```{r}

```

