---
title: "Extra stuff"
output: html_notebook
---

Non-essential, exploratory, not-in-pub plots and extra analyses can go here. Mostly just because I hate deleting stuff.
Main expression analysis script is eQTLdata_xyexpression.Rmd
Gene loss script in jpop_tx_exp_geneloss.Rmd


```{r}
library(tidyverse)
library(DESeq2)
library(gtools)

`%nin%` <- negate(`%in%`)

get_obs_table <- 
  function(data, cutoff) {
  data %>%
    filter(res.allele.padj < 0.1, abs(res.allele.log2FoldChange) > cutoff) %>%
    summarise(n_Y_oe = sum(res.allele.log2FoldChange > 0),
              n_X_oe = sum(res.allele.log2FoldChange <= 0)) %>%
    mutate(foldChangeCutoff = cutoff)
  }

reads_pg_pvals<-read_csv("data/rna_ase_results_eqtl_sept3.csv")
dnaCountsClean<-read_csv("data/dnaCountsClean.csv")
```


```{r}
ggplot(reads_pg_pvals) +
  geom_point(aes(x=start_tx_pat, 
                 y = 2^res.allele.log2FoldChange/(1+2^res.allele.log2FoldChange), size = res.allele.lfcSE,
                 color = sigTest),alpha = 0.5) 
#ggsave("lfc_ase_log2_1.pdf")
ggplot(reads_pg_pvals) +
  geom_point(aes(y=(maleCountsTotal.pat+maleCountsTotal.mat)/(femaleCountsTotal.mat+femaleCountsTotal.pat), 
                 x = maleCountsTotal.pat/(maleCountsTotal.mat),
                 color = sigTest)) +ylim(0,1) + xlim(0,1) +geom_abline(intercept = 0.5, slope = 0.5)
#ggsave("DC_expected_counts_Lfc1_axis.pdf")

# fraction of genes ow or ue on Y depending on stringency
# more stringent lgo2fc2 change, more reduced Y exp, but fewer sig sites
```

## exploratory dna counts plots - mismapping
### explore DNA output
There is some dna mapping to Y in females.

```{r}
ggplot(dnaCountsClean, aes(meanCounts))+geom_histogram() + facet_wrap(~ Chr)
ggplot(dnaCountsClean, aes(female_mean))+geom_histogram() + facet_wrap(~ Chr)
#ggsave("femaleMeanCounts_nolim_hist.pdf")
```
If we filter mean counts <200, we can get a closer look at what's going on. There are a high number of genes mapping to Y in the females, the mean mapping is somewhat lower than the other chromosomes. 

```{r}
ggplot(dnaCountsClean, aes(female_mean))+geom_histogram() + xlim(0,200)+ facet_wrap(~ Chr)
#ggsave("femaleMeanCounts_lim200_hist.pdf")
```

Filtering out mean pat reads <20 should remove a good chunk of problems? but many problematic genes remain. Whats the tolerable lower bound?
```{r}
dnaCountsClean %>% filter(Chr == "Y") %>% ggplot(aes(female_mean)) + geom_histogram(bins = 1000)+ 
  xlim(0,400)
dnaCountsClean %>% filter(Chr == "Y") %>% ggplot(aes(female_mean)) + geom_histogram(bins = 1000)+ 
  xlim(20,400)
```

```{r}
ggplot() +geom_histogram(data = goodgenes_dna,aes(female_mean.pat))
ggplot() +geom_histogram(data = goodgenes_dna,aes(female_mean.mat))

ggplot() +geom_histogram(data = goodgenes_dna,aes(male_mean.pat))

```

```{r}
ggplot(dnaCountsClean, aes(male_mean))+geom_histogram() + xlim(0,200)+ facet_wrap(~ Chr)
ggsave("maleMeanCounts_lim200_hist.pdf")
ggplot(dnaCountsClean, aes(male_mean))+geom_histogram() + facet_wrap(~ Chr)
ggsave("maleMeanCounts_nolim_hist.pdf")

```
