---
title: "PCAs. of genomic data to see which samples are possibly abberant/need removing"
output: html_notebook
---

males and females should form two diff clusters.
```{r}
library(tidyverse)
library(ggfortify)
library(ggrepel)
```

PCA of coverage DNA males and females
* note: the cleaning step has been tweaked
* samples 24fM should always be removed
* for main expression analyses, we are leaving out 40aF and 35aM - looks like opposite sex - mislabeling or abberrant
```{r} 
#library(factoextra) 

# dnamat<-dnaCountsClean %>% select(-Chr:-Length) %>% column_to_rownames(var = "Geneid") %>% 
#   select(-"24fM") 
#  #select(-"40aF",-"24fM",-"35aM") %>%
# dnamat_flip<- t(dnamat)
# ## pca error - there are some genes with zero variance which need to be filtered out!
# ## probbly zero reads or smoethig..
# which(apply(dnamat_flip, 2, var)==0)
# dnamat_flip<-dnamat_flip[ , which(apply(dnamat_flip, 2, var) != 0)]
# 
# # dnamat_flip_sex<-as.data.frame(dnamat_flip) %>% 
# #   rownames_to_column("Sample")  %>% 
# #   mutate(Sex = if_else(grepl("*M", Sample),"M", "F"))  %>% relocate(Sample,Sex)
# 
# 
# res.pca <- prcomp(dnamat_flip, scale = TRUE)
# 
# 
# autoplot(res.pca, data = filter(sampleInfo, Sample !="24fM"), colour="Sex")
# 
# ggsave("PCA_DNAcounts.pdf")

#autoplot(res.pca, data = dnamat_flip_sex, colour="Sex") + geom_label_repel(aes(label = Sample,color = Sex), max.overlaps = 30)

## abberant male is 35aM
#View(select(dnaCountsClean,Chr,"35aM","43aF","28dM","50bM"))
## need to be filtered out

```

```{r}
# ### pca if I remove the sex chromosomes???
# dnamat_nosex<-dnaCountsClean %>% select(-"24fM") %>%
#   filter(Chr != "X" & Chr != "Y") %>%select(-Chr:-Length) %>% column_to_rownames(var = "Geneid")
# 
# dnamat_flip_nosex<- t(dnamat_nosex)
# ## pca error - there are some genes with zero variance which need to be filtered out!
# ## probbly zero reads or smoethig..
# 
# which(apply(dnamat_flip_nosex, 2, var)==0)
# dnamat_flip_nosex<-dnamat_flip_nosex[ , which(apply(dnamat_flip_nosex, 2, var) != 0)]
# res.pca.nosex <- prcomp(dnamat_flip_nosex, scale = TRUE)
# 
# autoplot(res.pca.nosex, data = (filter(sampleInfo, Sample !="24fM")), colour="Sex") + ggrepel::geom_label_repel(aes(label = Sample))
# 
# ggsave("pca_dna_nosexchr.pdf")
# 
# 
# # norm coverage
# normFac<-as.data.frame(dnamat_flip_nosex) %>% 
#   rownames_to_column("Sample") %>% 
#   rowwise() %>%
#   mutate(TotalReads = sum(c_across(where(is.numeric)))) %>% 
#   relocate(TotalReads) 
# 
# normalizeDnaCounts<- normFac %>% mutate(across(where(is.numeric), ~ ./TotalReads)) %>% select(-TotalReads)
# 
# res.pca.nosex.norm<- prcomp((normalizeDnaCounts %>% column_to_rownames("Sample")), scale = TRUE) 
# 
# autoplot(res.pca.nosex.norm, data = (filter(sampleInfo, Sample !="24fM")), colour="Sex") + 
#   ggrepel::geom_label_repel(aes(label = Sample))
# 
# ggsave("pca_dna_nosexchr_norm.pdf")
# ## THHIS IS A RABBIT HOLE
# #MOVE ON!!!
```

