---
title: "dosage compensation - comparative"
output: html_notebook
---

How does expression of sagittatus orthologous genes compare to the X-Y gametologs in Rhast?

1. Run DESEQ!
```{r}
library(DESeq2)
library(tidyverse)
```

## Sag deseq
```{r}
rnaCountsSag<-read_table("readCountsSagittatus/sagittatusRnaCounts.txt")
rnaColnamesSag<-colnames(rnaCountsSag)
rnaColnamesRepairedSag <- gsub("(Rsg.+)Aligned.+", "\\1", rnaColnamesSag)
colnames(rnaCountsSag)<-rnaColnamesRepairedSag 

readsMatSag<-rnaCountsSag %>% column_to_rownames("Geneid") %>% select(RsgL1:Rsg_1_1) # convert to matrix of all genes and inds

sampleNamesSag<- data.frame(colnames(readsMatSag)) 

```


```{r}
## WAIT did I want to normalize these? What am I doing really?
#sagOrths<-read_csv("hyphy/orthsUpdatedAnnos.csv.gz")%>%
#  mutate(across(everything(),~str_replace(.,"-RA$","")))
#orths<-read_delim("data/Rhast_TX_pangenes.txt.gz")

pgs<-data.table::fread("data/wide_rhast_tx_sag_pangenes.txt.gz") 
pgs_filt<- pgs %>% filter(Rhast_TX !="" & Rsag !="") %>% filter(!grepl("\\|",Rhast_TX) & !grepl("\\|",Rsag)) %>% 
  mutate(orthType=case_when(grepl("\\*",Rsag) ~ "NSOrth",
                            grepl("\\*",Rhast_TX) ~ "NSOrth",
                            TRUE ~ "PASS")) %>%
     mutate(Rsag = str_remove(Rsag, "-RA\\*?$"),
            Rhast_TX = str_remove(Rhast_TX, "-RA\\*?$")) %>%  
  dplyr::arrange(desc(orthType)) %>%
  distinct(Rhast_TX, .keep_all = TRUE) %>%
  distinct(Rsag, .keep_all =TRUE) %>%
  unite(pairID, c("Rhast_TX", "Rsag"),remove = FALSE)
  
```


```{r}
#tx_yloss_nc_overlap<-read.csv("data/tx_yloss_overlapNCcomplete_loss.csv") 
tx_yloss<-read_csv("data/tx_yloss_blastconfirmed_Aug2024.csv")


```

```{r}
reads_pg_pvals<-read_csv("data/rna_ase_results_eqtl_sept3.csv.gz",show_col_types =FALSE)

x_overexpressed<-filter(reads_pg_pvals, res.allele.padj < 0.1 & res.allele.log2FoldChange < 1) %>%
  select(id_tx_mat, id_tx_pat)

y_overexpressed<-filter(reads_pg_pvals, res.allele.padj < 0.1 & res.allele.log2FoldChange > 1) %>% select(id_tx_mat,id_tx_pat)

```

```{r}
eQTL_counts<-read_delim("readCountsRNAeQTL/rhastTXeQTLrna.txt") #%>%
  select(!2:7)

sag_counts<-read_delim("readCountsSagittatus/sagittatusRnaCounts.txt") #%>%
  select(!2:7)

rnaColnames<-colnames(eQTL_counts)
rnaColnamesRepaired <- gsub(".*NEBNext_dual_i._.+\\.([0-9]+[a-z]+[FM])LRAligned.+", "\\1", rnaColnames)
colnames(eQTL_counts)<-rnaColnamesRepaired 

sagColnames<-colnames(sag_counts)
sagColnamesRepaired<-gsub("Aligned\\.sortedByCoord\\.out","\\1",sagColnames)
colnames(sag_counts)<-sagColnamesRepaired 

geneNames<-pgs_filt %>% select(pairID,Rhast_TX,Rsag,chr)

allReadCounts<-left_join(geneNames,eQTL_counts, by = c("Rhast_TX" = "Geneid")) %>%
  left_join(.,sag_counts,by = c("Rsag" = "Geneid")) %>% column_to_rownames(var = "pairID") #%>% select(!c("Rhast_TX","Rsag","chr"))

sampleNames<- colnames(allReadCounts)
sampleInfo<-data.frame(sampleNames) 
# remove headers


df_all_sag<-DESeqDataSetFromMatrix(allReadCounts,sampleInfo,design = ~ 1)
dds_all <- DESeq(df_all)
allchrFactors<-sizeFactors(dds_all)
normCounts<-counts(dds_all, normalized=TRUE)
colnames(normCounts)

write_delim(as.data.frame(normCounts),"data/normCounts_sag_and_rhast.txt")


```

```{r}
normcounts_clean<-as.data.frame(normCounts) %>% 
  rownames_to_column(var = "pairID") %>%
  rowwise %>% 
  dplyr::mutate(maleMean = mean(c_across(contains("M",ignore.case = FALSE) ))) %>%
  dplyr::mutate(maleVar = sd(c_across(contains("M",ignore.case = FALSE)))) %>%
  dplyr::mutate(femaleMean = mean(c_across(contains("F",ignore.case = FALSE)))) %>%
  dplyr::mutate(femaleVar = sd(c_across(contains("F",ignore.case = FALSE)))) %>%
  dplyr::mutate(totalCount =sum(c_across("39dF":"Rsg_1_1"))) %>%
  dplyr::mutate(totalVar = sd(c_across("39dF":"Rsg_1_1"))) %>%
  dplyr::mutate(sagMean = mean(c_across(contains("Rsg",ignore.case = FALSE)))) %>%
  dplyr::mutate(sagVar = sd(c_across(contains("Rsg",ignore.case = FALSE)))) %>%
  ungroup() %>%
  filter(totalCount >20) 

normcounts_genes<-left_join(normcounts_clean,geneNames,by = "pairID") %>% relocate(pairID,Rhast_TX,Rsag)
  
dc_combo <- normcounts_genes %>% mutate(genetype = case_when(
   #Geneid%in%completely_silenced$id_tx_mat ~ "completeYsilenced",
    Rhast_TX%in%x_overexpressed$id_tx_mat ~ "xOverexp",
    Rhast_TX%in%y_overexpressed$id_tx_pat ~ "yOverexp",
    Rhast_TX%in%tx_yloss_nc_overlap$Geneid ~ "hemizygous", TRUE ~ "other")) %>% 
    filter(genetype != "other")


plot1<-dc_combo %>% filter(chr =="X") %>% ggplot(aes(x = femaleMean, y = sagMean)) + geom_point() + 
  geom_abline(intercept = 0, slope = 1) + theme_bw() + scale_y_continuous(limits = c(0,5000)) + scale_x_continuous(limits = c(0,6000))+
  geom_smooth(method = "lm")
plot2<-dc_combo %>% filter(chr =="X") %>% ggplot(aes(x = maleMean, y = sagMean)) + geom_point() + theme_bw() +
  geom_abline(intercept = 0, slope = 1) + scale_y_continuous(limits = c(0,5000)) + scale_x_continuous(limits = c(0,6000)) +
    geom_smooth(method = "lm")

cowplot::plot_grid(plot1,plot2) 
## sag expression is 
ggsave("compare_sag_rhast_mf.pdf",w = 8,h =6)


```

```{r}

```


