---
title: "expression analyses with eQTL data"
output: html_notebook
---

#### Packages & such
```{r packages, echo=TRUE, message=FALSE, warning=FALSE}
library(tidyverse)
library(DESeq2)
#source("gametolologExpressionDataProcessing.R")
```


```{r}
reads_pg_pvals_dna<- read_csv("data/dna_ase_results_eqtl_noPAR.csv") # read in data
setwd(rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(
    # This should allow Rmarkdown to locate the data
    root.dir = rprojroot::find_rstudio_root_file()
)
```



#### Expression analyses
RNA read counts.
```{r}
badgenes<-rnaCountsClean %>% 
  filter(female_mean>= 20 & Chr =="Y") %>% select(Geneid) # get genes where females have too much Y coverage
# some may be on Y PAR - which is fromm 0-45MB. Disregard and remove all for now!
badgenes_x <- pgMatOrths %>% select(id_tx_mat,id_tx_pat) %>% filter(id_tx_pat%in%badgenes$Geneid) %>% select(id_tx_mat)
## find corresponding X gene for every "bad" Y gene
shared_bad_biased<-intersect(badgenes$Geneid,biasedgenes_dna$id_tx_pat) # 30 genes
shared_bad<-intersect(badgenes$Geneid,badgenes_dna$id_tx_pat) # 394 genes
```


```{r}
readsMatAll<-rnaCountsClean %>% column_to_rownames("Geneid") %>% select("39dF":160) # convert to matrix of all genes and inds

readsMatAuto<-rnaCountsClean %>% filter(grepl("A",Chr)) %>% # matrix with just autosomes, for generating norm factors
   column_to_rownames("Geneid") %>% select("39dF":160) 

sampleNames<- colnames(readsMatAll)
#
sampleInfo<-data.frame(sampleNames) %>% 
  mutate(.,Sex = if_else(grepl("*M", sampleNames),"M", "F"))

df_all<-DESeqDataSetFromMatrix(readsMatAll,sampleInfo,design = ~ 1)
df_auto<-DESeqDataSetFromMatrix(readsMatAuto,sampleInfo,design = ~ 1)
dds_all <- DESeq(df_all)
dds_auto <- DESeq(df_auto)
allchrFactors<-sizeFactors(dds_all)
autoFactors<-sizeFactors(dds_auto)
compareFactors<-data.frame(allchrFactors,autoFactors) %>% rownames_to_column("sample") 
compareFactorsLong <- compareFactors %>% pivot_longer(cols = -sample, names_to = "normType", values_to = "value")
ggplot(compareFactorsLong) + geom_point(aes(x = sample, y = value, color = normType)) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

#ggsave("compare_norm_factors.pdf", h = 5, w = 18)
## very similar
## use autosome factors anyway *for dc results


```
Gene loss
- using blast confirmed gene loss, not just the overlapping with NC ones!
- need to record numbres of each though
```{r}
# load set of hemizygous genes, generated from jpop_tx_exp_geneloss.Rmd
tx_yloss<-read_csv("data/genelist_tx_mat.txt",col_names = FALSE)
tx_yloss_nc_overlap<-read.csv("data/tx_yloss_overlapNCcomplete_loss.csv")
hemiz_blast_confirm<-read.csv("data/tx_yloss_blastconfirmed_Aug2024.csv")
# 376 genes lost before blast filtering
intersect(tx_yloss_nc_overlap$repGene, hemiz_blast_confirm$tx_mat) %>% length()
intersect(hemiz_blast_confirm$tx_mat, tx_yloss_nc_overlap$repGene) %>% length()
# all 125 nc overlap lost genes are also confirmed to be either completely or partially lost via blast
hemiz_blast_partial<-read.csv("data/tx_yloss_blastconfirmed_partial_Aug2024.csv")
hemiz_blast_complete<-read.csv("data/tx_yloss_blastconfirmed_complete_Aug2024.csv")

# overlaps once more
intersect(hemiz_blast_complete$tx_mat, tx_yloss_nc_overlap$repGene) %>% length() # 112
intersect(hemiz_blast_partial$tx_mat, tx_yloss_nc_overlap$repGene) %>% length() # 13
setdiff(hemiz_blast_complete$tx_mat, tx_yloss_nc_overlap$repGene) %>% length() #110/222 completely lost genes are not lost in NC
setdiff(hemiz_blast_partial$tx_mat, tx_yloss_nc_overlap$repGene) %>% length() # 117/130 partially lost genes are not lost in NC

#creating a data frame using dplyr where one column is gene name, the other is category (e.g. complete loss, partial loss), plus a third column for overlap with NC gene loss

geneloss_categorized <- hemiz_blast_confirm %>% mutate(category = case_when(
  tx_mat %in% hemiz_blast_complete$tx_mat ~ "complete loss",
  tx_mat %in% hemiz_blast_partial$tx_mat ~ "partial loss",
  TRUE ~ "no loss")) %>% 
  mutate(nc_overlap = ifelse(tx_mat %in% tx_yloss_nc_overlap$repGene, "yes", "no"))

test1<-chisq.test(table(geneloss_categorized$category,geneloss_categorized$nc_overlap))
#X-squared = 56.827, df = 1, p-value = 4.759e-14

geneloss_summary<- geneloss_categorized %>% group_by(category,nc_overlap) %>% summarise(n = n())              
```

```{r}

# replace og coutn tabels size factors
dds_all$sizeFactor<-autoFactors
dds_all$sizeFactor

normalized_counts<-counts(dds_all, normalized=TRUE)
#normcounts<-as.data.frame(normalized_counts) %>% rownames_to_column()
#write_csv(normcounts,"norm_counts_autosomefactors_eQTLleafRhast.csv")
normcounts<-read_csv("data/norm_counts_autosomefactors_eQTLleafRhast.csv.gz")
hemicounts <- as.data.frame(normalized_counts) %>% 
  rownames_to_column(var = "Geneid") %>% 
  filter(Geneid%in%hemiz_blast_confirm$tx_mat) %>%
  mutate(category = case_when(
    Geneid %in% hemiz_blast_complete$tx_mat ~ "complete loss",
    Geneid %in% hemiz_blast_partial$tx_mat ~ "partial loss",
      TRUE ~ "no loss")) %>% 
  mutate(nc_overlap = ifelse(Geneid %in% tx_yloss_nc_overlap$repGene, "yes", "no")) %>%
  rowwise %>% 
  mutate(maleMean = mean(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(maleVar = sd(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(maleSE = plotrix::std.error(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(femaleMean = mean(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(femaleVar = sd(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(femaleSE = plotrix::std.error(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(totalCount =sum(c_across("39dF":"43bM"))) %>%
  mutate(totalVar = sd(c_across("39dF":"43bM"))) %>%
  ungroup() %>%
  filter(totalCount >20)

# this plot shows SE error bars. Yeesh!!!!
ggplot(hemicounts, aes(x = (femaleMean), y = (maleMean))) + 
  geom_point() +
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE)) +
  #geom_abline(intercept = coef(fit)[1], slope = coef(fit)[2], col='red') +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  #scale_x_continuous(trans ="log2") +#expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) + 
  #scale_y_continuous(trans = "log2") +#expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) +
  xlab("Average female expression") +
  ylab("Average male expression")
ggsave("plots/hemizygous_dosage_SE.png", h = 5, w = 7)

### sooo much less shitty!
### zoom in 

ggplot(hemicounts, aes(x = (femaleMean), y = (maleMean))) + 
  geom_point() +
  ylim(0,2000) +
  xlim(0,2000) +
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE)) +
  geom_smooth(method="lm", se =TRUE) +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  xlab("Average female expression") +
  ylab("Average male expression")
# notice that a few genes have some weird evidence of DC stuff? possibly??
ggsave("plots/hemizygous_dosage_lim2000.png", h = 5, w = 7)
# zoom out
ggplot(hemicounts, aes(x = (femaleMean), y = (maleMean))) + 
  geom_point() +
  ylim(2000,10000) +
  xlim(2000,20000) +
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE)) +
  geom_smooth(method="lm", se = TRUE) +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  xlab("Average female expression") +
  ylab("Average male expression")
ggsave("plots/hemizygous_dosage_largeVals.png", h = 5, w = 7)


hemi_lm<- lm(maleMean ~ femaleMean, data = hemicounts)
summary(hemi_lm)
# slope =. 0.450908
# Adjusted R-squared:  0.9939 
# p-value: < 2.2e-16

# 0.45 is the slope of the line. 0.5 is the expected slope for NO dosage compensation
# 1 is the expected slope for dosage compensation. 



```
Same plots but add colors/shapes for different types of hemizygous genes
```{r}
ggplot(hemicounts, aes(x = (femaleMean), y = (maleMean))) + 
  ylim(0,2000) +
  xlim(0,2000) +
  geom_point(aes(colour = category)) +
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE), alpha = 0.5) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE),alpha = 0.5) +
  #geom_abline(intercept = coef(fit)[1], slope = coef(fit)[2], col='red') +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  #scale_x_continuous(trans ="log2") +#expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) + 
  #scale_y_continuous(trans = "log2") +#expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) +
  xlab("Average female expression") +
  ylab("Average male expression")
ggsave("plots/hemizygous_dosage_SE_cat_lim2000.png", h = 5, w = 7)

# now coloring by whether there is overlap with NC
ggplot(hemicounts, aes(x = (femaleMean), y = (maleMean))) + 
  ylim(0,2000) +
  xlim(0,2000) +
  geom_point(aes(colour = nc_overlap)) +
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE), alpha = 0.5) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE),alpha = 0.5) +
  #geom_abline(intercept = coef(fit)[1], slope = coef(fit)[2], col='red') +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  xlab("Average female expression") +
  ylab("Average male expression")
ggsave("plots/hemizygous_dosage_SE_overlap_lim2000.png", h = 5, w = 7)

ggplot(hemicounts, aes(x = (femaleMean), y = (maleMean))) + 
  geom_point(aes(colour = nc_overlap)) +
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE), alpha = 0.5) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE),alpha = 0.5) +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
   xlab("Average female expression") +
  ylab("Average male expression")

# plot hist of female/male expression
ggplot(hemicounts) +
  geom_histogram(aes(x=maleMean/femaleMean, colour = nc_overlap),position="stack") +
  xlim(0,2)

# chisquare to see if there are more or less of the diff types of genes in each group of male/female X expression!?

breaks <- c(-Inf, 0.2, 0.4, 0.6, 0.8, 1, 1.2, 1.4, 1.6, 1.8, 2, 2.2, 2.4, Inf)
labels <- c("0-0.2", "0.2-0.4", "0.4-0.6", "0.6-0.8","0.8-1.0", "1.0-1.2", "1.2-1.4", "1.4-1.6", "1.6-1.8","1.8-2.0","2.0-2.2","2.2-2.4",">2.4")


dosageSummaryCat <- hemicounts %>% 
  mutate(dosageRatio = maleMean/femaleMean) %>% 
  mutate(dc_range = cut(dosageRatio, breaks = breaks, labels = labels, right = FALSE)) %>%
  group_by(category,dc_range) %>% summarise(n = n()) 

dosageSummaryOverlap <- hemicounts %>% 
  mutate(dosageRatio = maleMean/femaleMean) %>% 
  mutate(dc_range = cut(dosageRatio, breaks = breaks, labels = labels, right = FALSE)) %>%
  group_by(nc_overlap,dc_range) %>% summarise(n = n()) 

# highly doubt there's any interesting pattern here LOL
# was fun to check I guess

```


## ASE


```{r}
eqtl_x<- filter(rnaCountsClean, Chr =="X") %>% select(c(Geneid,totalCountsGene:"43bM")) ## extract X gene counts
eqtl_y<- filter(rnaCountsClean, Chr =="Y") %>% select(c(Geneid,totalCountsGene:"43bM")) ## extract Y gene counts

reads_pg_nofilter<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>%
  drop_na()%>%  summarise(n = n())

reads_pg_PARfilter<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>%
  filter(start_tx_pat > 45000000) %>% summarise(n = n())

reads_pg_badRNAgenesFilter<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>%
  filter(id_tx_mat%nin%badgenes_x$id_tx_mat) %>% # weird expression filter
  filter(id_tx_pat%nin%badgenes$Geneid) %>%
  filter(start_tx_pat > 45000000) %>%
  summarise(n = n())

reads_pg_badDNAgenesFilter<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>%
  filter(id_tx_mat%nin%badgenes_x$id_tx_mat) %>% # weird expression filter
  filter(id_tx_pat%nin%badgenes$Geneid) %>%
  filter(start_tx_pat > 45000000) %>%
  #filter(pairID%nin%biasedgenes_dna$pairID) %>% # biased gene filter
  filter(pairID%nin%badgenes_dna$pairID) %>%
  summarise(n = n())

reads_pg_biasedDNAgenesFilter<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>%
  filter(id_tx_mat%nin%badgenes_x$id_tx_mat) %>% # weird expression filter
  filter(id_tx_pat%nin%badgenes$Geneid) %>%
  filter(start_tx_pat > 45000000) %>%
  filter(pairID%nin%biasedgenes_dna$pairID) %>% # biased gene filter
  filter(pairID%nin%badgenes_dna$pairID) %>%
  summarise(n = n())
 
reads_pg<-left_join(pgMatOrths,eqtl_x, by = c("id_tx_mat"="Geneid")) %>%
  left_join(.,eqtl_y, by = c("id_tx_pat" = "Geneid"),suffix = c(".mat",".pat")) %>% 
  filter(id_tx_mat%nin%badgenes_x$id_tx_mat) %>% # weird expression filter
  filter(id_tx_pat%nin%badgenes$Geneid) %>%
  filter(start_tx_pat > 45000000) %>% # par filter
  filter(pairID%nin%biasedgenes_dna$pairID) %>% # biased gene filter
  filter(pairID%nin%badgenes_dna$pairID) %>% # filter weird Y mapping
  drop_na() %>% select(-flag_tx_pat,-flag_tx_mat) %>% 
  unique() %>%
  rowwise() %>% 
  mutate(totalCountAll=(totalCountsGene.mat+totalCountsGene.pat)) %>%
  ungroup() %>% filter(totalCountAll >20)

reads_pg_allfilters<-reads_pg %>% summarise(n=n())
# 274 genes remain
```
  

```{r}
# table of read numbers
category<-c("reads_pg_nofilter","reads_pg_PARfilter","reads_pg_badRNAgenesFilter","reads_pg_badDNAgenesFilter","reads_pg_biasedDNAgenesFilter","reads_pg_allfilters")
counts<-c(reads_pg_nofilter$n,reads_pg_PARfilter$n,reads_pg_badRNAgenesFilter$n,reads_pg_badDNAgenesFilter$n,reads_pg_biasedDNAgenesFilter$n,reads_pg_allfilters$n)

filtering_summary<-data.frame(category,counts)
print(filtering_summary)

ggplot(reads_pg,aes(x=male_mean.mat,y=male_mean.pat)) + geom_point() +
    geom_abline(slope = 1, intercept = 0, linetype=1) +
  theme_bw()+
  geom_smooth(method ="lm") +
  xlab("mean exp of X gametolog") +
  ylab("mean exp of Y gametolog") +
  scale_x_continuous(transform = "log2") +
  scale_y_continuous(transform = "log2")
ggsave("gametolog_readCounts_all_may27_logscale.jpg", h = 5, w = 7)

```
deseq ase test
```{r}
ase_data<-reads_pg %>% select(c(pairID,contains("M", ignore.case = FALSE ))) %>% column_to_rownames(var ="pairID")
## ase matrix
IDs<-colnames(ase_data)
library(gtools)
sorted_col_names <- mixedsort(IDs)
ase_data <- ase_data[, sorted_col_names]
#colnames(ase_data)
# yay!!
sample<-str_remove(sorted_col_names,"..at") # tm1 tm1 etc.
allele<-str_extract(sorted_col_names,".at$") # mat pat mat pat

colInfo<-data.frame(sorted_col_names,sample,allele) # wooo

ase_matrix<-as.matrix(ase_data)
design<- ~sample+allele

m <- 75

dds <- DESeqDataSetFromMatrix(ase_data, colInfo, design)
sizeFactors(dds) <- rep(1, 2*m)
dds <- DESeq(dds, fitType="local")

resultsNames(dds)

res.allele <- results(dds, name="allele_pat_vs_mat")
head(res.allele$log2FoldChange)

adj_pvals<-res.allele$padj


pvals_genes<-data.frame(res.allele@rownames,res.allele$padj,res.allele$log2FoldChange,res.allele$lfcSE) #%>%
reads_pg_pvals<-inner_join(reads_pg, pvals_genes, by =c("pairID" ="res.allele.rownames"))%>%
  mutate(sigTest = # just do significance, leave the log2foldchane to be filtered later omg
           case_when(res.allele.padj < 0.1 ~ "sig", #& abs(res.allele.log2FoldChange) > 1 ~ "sig",
                     TRUE ~"nonsig" )) %>% 
  relocate(c(pairID:male_mean.mat, meanCounts.pat:male_mean.pat, res.allele.padj:sigTest))

#write_csv(reads_pg_pvals, "data/rna_ase_results_eqtl_sept3.csv")
#reads_pg_pvals<-read_csv("data/rna_ase_results_eqtl_may17.csv")
reads_pg_pvals<-read_csv("data/rna_ase_results_eqtl_sept3.csv.gz")

head(reads_pg_pvals)
```


### read ratio, not using fold change ratio

```{r}
#df<- reads_pg_pvals %>% mutate(readratio = (male_mean.pat/male_mean.mat))

#ggplot(df)+ geom_histogram(aes(x = (readratio), color = sigTest)) #+xlim(0,1)
reads_pg_pvals2 <- reads_pg_pvals %>%   mutate(sigLFC = 
           case_when(res.allele.padj < 0.1 & abs(res.allele.log2FoldChange) > 1 ~ "True",
                     TRUE ~"False" ))
reads_pg_pvals2$sigLFC<-factor(reads_pg_pvals2$sigLFC, levels = c("True","False"))
mycolors<-RColorBrewer::brewer.pal(2,"Set1")[2:3]

ggplot(reads_pg_pvals2,aes(x=male_mean.mat,y=male_mean.pat, color = sigLFC)) + geom_point(aes(size = res.allele.lfcSE),alpha =0.60) +
    geom_abline(slope = 1, intercept = 0, linetype=1) +
  theme_bw()+
  geom_smooth(method ="lm") +
  xlab("mean expression of X\n gametolog in males") +
  ylab("mean expression of Y\n gametolog in males") +
  scale_x_continuous(transform = "log2",limits = c(NA,4000),breaks = c(0.0625,2,64,2048)) +
  scale_y_continuous(transform = "log2",limits = c(NA,4000),breaks = c(0.0625,2,64,2048)) +
  labs(color= "p-val < 0.1 & |log2FC| >1 ", size = "log2FC std. error") +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        legend.text = element_text(size =12),
        legend.title = element_text(size =12)) 
  
 #scale_color_manual(values = mycolors)
  #ggsave("gametolog_expression_lfc1_fdr0.1.pdf")
#ggsave("gametolog_expression_lfc1_fdr0.1_lfcSE.pdf")
ggsave("plots/gametolog_expression_lfc1_fdr0.1_lfcSE_logscale.pdf",h =6, w=10)


```

```{r}
ggplot(filter(reads_pg_pvals, 
              sigTest == "sig", res.allele.log2FoldChange < 0),
       aes(x=female_mean.mat, y=male_mean.mat)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_smooth(method ="lm") +
   #scale_x_continuous(expand = c(0, 0), limits = c(0, NA) ,breaks =seq(0,16000,2000)) + 
  #scale_y_continuous(expand = c(0, 0), limits = c(0, NA),breaks =seq(0,16000,2000)) +
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic()
### barely any sig ase??

ggplot(filter(reads_pg_pvals, res.allele.log2FoldChange > 0),
       aes(x=female_mean.mat, y=male_mean.mat)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_smooth(method ="lm") +
   #scale_x_continuous(expand = c(0, 0), limits = c(0, NA) ,breaks =seq(0,16000,2000)) + 
  #scale_y_continuous(expand = c(0, 0), limits = c(0, NA),breaks =seq(0,16000,2000)) +
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic()
### add error bar,
# stic
```
nearly done!
- add DNA
- sig testing?
- outgroup
- Ks vs y/x
- blast tx missing genes to tx genome????
  - NC shared set is likely conservative and good to go!
  - haven't id'd partial loss..
- non-one-one orthologs - explore 
- characterise these other genes
- GO GO enrichment

Need table
- how many Y oe vs. X oe genes for each fold change from 0.5 -> 1 -> 2 ... 
- at alpha of FDR 0.1
```{r}



fold_change_cutoffs <- c(0.5, 1, 1.5, 2)
obs_tables <- lapply(fold_change_cutoffs, function(cutoff) {
  get_obs_table((filter(reads_pg_pvals, res.allele.padj < 0.1)), cutoff)
})

combined_table <- do.call(rbind, obs_tables)
print(combined_table)
write_csv(combined_table, "n_xy_oe_RNA_eQTL_lfcRanges.csv")


```
## dosage compensation in "silenced" genes

```{r}
### silenced-ish genes
silenced_set<-filter(reads_pg_pvals, res.allele.padj < 0.1 & res.allele.log2FoldChange < 1) %>% select(id_tx_mat)

silenced_counts <- as.data.frame(normalized_counts) %>% 
  rownames_to_column(var = "Geneid") %>% filter(Geneid%in%silenced_set$id_tx_mat) %>%
  rowwise %>% 
  mutate(maleMean = mean(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(maleVar = sd(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(maleSE = plotrix::std.error(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(femaleMean = mean(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(femaleVar = sd(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(femaleSE = plotrix::std.error(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(totalCount =sum(c_across("39dF":"43bM"))) %>%
  mutate(totalVar = sd(c_across("39dF":"43bM"))) %>%
  ungroup() %>%
  filter(totalCount >20)

 
ggplot(silenced_counts, aes(x = femaleMean, y = maleMean)) + 
  geom_point() +
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE)) +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  #scale_x_continuous(expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) + 
  #scale_y_continuous(expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) +
  xlab("Average female expression") +
  ylab("Average male expression")

ggsave("silenced_yunderexpressed_dosagecomp.jpg",w = 7, h = 5)


ggplot(silenced_counts, aes(x = femaleMean, y = maleMean)) + 
  geom_point() +
  ylim(0,100) +
  xlim(0,100)+
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE)) +
  geom_smooth(method="lm") +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +theme_classic() +
  #scale_x_continuous(expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) + 
  #scale_y_continuous(expand = c(0, 0), limits = c(0, 26000),breaks =seq(0,24000,2000)) +
  xlab("Average female expression") +
  ylab("Average male expression")
ggsave("silenced_yunderexpressed_dosagecomp_100limits.jpg",w = 7, h = 5)

```

Dosage compensation plot combined
```{r}
#completely_silenced<-filter(reads_pg_pvals, res.allele.padj < 0.1 & res.allele.log2FoldChange < 1 & male_mean.pat == 0) %>%
#  select(id_tx_mat)
x_overexpressed<-filter(reads_pg_pvals, res.allele.padj < 0.1 & res.allele.log2FoldChange < 1) %>%
  select(id_tx_mat, id_tx_pat)
y_overexpressed<-filter(reads_pg_pvals, res.allele.padj < 0.1 & res.allele.log2FoldChange > 1) %>% select(id_tx_mat,id_tx_pat)


normcounts_clean<-#as.data.frame(normalized_counts) %>% 
  normcounts %>%
  dplyr::rename("Geneid" ="rowname") %>%
  rowwise %>% 
  mutate(maleMean = mean(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(maleVar = sd(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(maleSE = plotrix::std.error(c_across(contains("M",ignore.case = FALSE)))) %>%
  mutate(femaleMean = mean(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(femaleVar = sd(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(femaleSE = plotrix::std.error(c_across(contains("F",ignore.case = FALSE)))) %>%
  mutate(totalCount =sum(c_across("39dF":"43bM"))) %>%
  mutate(totalVar = sd(c_across("39dF":"43bM"))) %>%
  ungroup() %>%
  filter(totalCount >20) #%>%
dc_combo <- normcounts_clean %>% mutate(genetype = case_when(
   #Geneid%in%completely_silenced$id_tx_mat ~ "completeYsilenced",
    Geneid%in%x_overexpressed$id_tx_mat ~ "xOverexp",
    Geneid%in%y_overexpressed$id_tx_mat ~ "yOverexp",
    Geneid%in%hemiz_blast_confirm$tx_mat ~ "hemizygous", TRUE ~ "other")) %>% 
    filter(genetype != "other")


ggplot(dc_combo, aes(x = femaleMean, y = maleMean, color = genetype)) + 
  geom_point(alpha = 0.5, size = 3) +
  ylim(0,500)+
  xlim(0,500)+
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE, color = genetype)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE, color = genetype)) +
  geom_smooth(method="lm", fullrange = TRUE) +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +
  theme_bw() +
  xlab("Average female expression") +
  ylab("Average male expression") +
  labs(color= NULL) +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        legend.text= element_text(size =12)) +
  scale_color_discrete(labels = c("hemizygous"="Hemizygous", "xOverexp" = "Partial or complete\n Y silencing"))

#ggsave("plots/combined_dc_plot_hemi_ysilence_updated.png", h = 6, w =9)
ggsave("plots/combined_dc_plot_hemi_ysilence_limit500.png", h = 6, w =9)


ggplot(dc_combo, aes(x = femaleMean, y = maleMean, color = genetype)) + 
  geom_point(alpha = 0.5, size = 3) +
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE, color = genetype)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE, color = genetype)) +
  geom_smooth(method="lm", fullrange = TRUE) +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +
  theme_bw() +
  xlab("Average female expression") +
  ylab("Average male expression") +
  labs(color= NULL) +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        legend.text= element_text(size =12)) +
  scale_color_discrete(labels = c("hemizygous"="Hemizygous", "xOverexp" = "Partial or complete\n Y silencing")) +
  facet_wrap(~genetype, scales = "free")

```
Plot for the pub
- one w silenced genes and hamiz
- one for sup w y oveexp

```{r}
dc_combo %>% filter(genetype == "yOverexp") %>% 
  ggplot(aes(x = femaleMean, y = maleMean)) + 
  geom_point(size = 2) +
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE)) +
  geom_smooth(method="lm", fullrange = TRUE) +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +
  theme_bw() +
  xlab("Average female expression") +
  ylab("Average male expression") +
  labs(color= NULL) +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        legend.text= element_text(size =12))

ggsave("plots/yOverexp_dc_plot.png", h = 5, w = 7)

dc_combo %>% filter(genetype != "yOverexp") %>% ggplot(aes(x = femaleMean, y = maleMean, color = genetype)) + 
  geom_point(alpha = 0.5, size = 3) +
  geom_errorbar(aes(ymin = maleMean-maleSE, ymax = maleMean+maleSE, color = genetype)) +
  geom_errorbarh(aes(xmin = femaleMean-femaleSE, xmax = femaleMean+femaleSE, color = genetype)) +
  geom_smooth(method="lm", fullrange = TRUE) +
  geom_abline(slope = 1, intercept = 0, linetype=1) + 
  geom_abline(slope = 0.5, intercept = 0, linetype=2) +
  theme_bw() +
  xlab("Average female expression") +
  ylab("Average male expression") +
  labs(color= NULL) +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        legend.text= element_text(size =12)) +
  scale_color_discrete(labels = c("hemizygous"="Hemizygous", "xOverexp" = "Partial or complete\n Y silencing"))
ggsave("plots/hemiz_xOverexp_dc_plot.png", h = 5, w = 7)

```


Messing around w KS plots
```{r}
ks_raw<-read_table("data/Rhastxy.april2024.stephen.coge.aligncoords.gcoords.ks.txt", skip = 2) 
ks_colnames<-colnames(ks_raw)
ks_raw<-read_table("data/Rhastxy.april2024.stephen.coge.aligncoords.gcoords.ks.txt", comment = "#", col_names = ks_colnames)
mydelim = "\\|\\|"
col_parts_1 <- as.vector(unlist(str_split(colnames(ks_raw)[4],  mydelim)))
col_parts_2<- as.vector(unlist(str_split(colnames(ks_raw)[8],  mydelim)))


# Apply the function to the dataframe
ks_data_separated <- ks_raw %>% separate(
  col = `chr1||start1||stop1||name1||strand1||type1||db_feature_id1||genome_order1||percent_id1`, 
                                         sep = mydelim, into = col_parts_1) %>%
  separate(col = `chr2||start2||stop2||name2||strand2||type2||db_feature_id2||genome_order2||percent_id2`,
           sep = mydelim, into = col_parts_2) %>% select(-"b<db_genome_id>_<chr>",-"a<db_genome_id>_<chr>",-GEVO_link) %>%
  dplyr::rename("Ks"="#Ks") %>% mutate(id_tx_pat = gsub("-RA","",name1)) %>% mutate(id_tx_mat = gsub("-RA","",name2))
# View the modified dataframe
colnames(ks_data_separated)

### combine with y x exp data

reads_pg_pvals_ks <-left_join(reads_pg_pvals, ks_data_separated, by = c("id_tx_mat","id_tx_pat")) %>% 
  relocate(Ks:block_score) %>%
  drop_na() filter

ggplot(reads_pg_pvals_ks) + 
         geom_point(aes(y = Ks, x = male_mean.pat/male_mean.mat)) +
         theme_classic() + xlim(0,2) +ylim(0,0.3) 
ggsave("Ks_readRatio_filter.pdf")
ggsave("Ks_readTa")
ggplot(reads_pg_pvals_ks) + geom_point(aes(y = Ks, x = res.allele.log2FoldChange)) #+ ylim(0,0.5)  +xlim(-10,10)

```
- sliding windows?
- color code by site type (silenced etc.)
```{r}
ks_data_types<- ks_data_separated %>% mutate_at(c('start1', 'start2'), as.numeric) %>%
     mutate(genetype = case_when(
    (id_tx_mat%in%x_overexpressed$id_tx_mat & id_tx_pat%in%x_overexpressed$id_tx_pat) ~ "xOverexp",
    (id_tx_mat%in%y_overexpressed$id_tx_mat & id_tx_pat%in%y_overexpressed$id_tx_pat) ~ "yOverexp", TRUE ~ "other")) %>%  
    filter(Ks < 0.3) %>% filter(chr1  =="Y" & chr2 =="X")

ggplot(ks_data_types) + geom_point(aes(y = Ks, x = start2, color = genetype)) + 
  scale_color_manual(values=alpha(c("grey", "blue", "green"),c(0.2, 0.5, 0.5))) + ylab("Ks between X and Y genes") + xlab("gene start position on X chromosome") +
  theme_classic() + geom_vline(xintercept = 220e6, linetype = "dashed")# scale_x_continuous(breaks = seq(0,500e6,100e6)) +theme_classic()
ggsave("Ks_labels_xcoord.jpg",h=5,w=7)

ggplot(ks_data_types) + geom_point(aes(y = Ks, x = start1, color = genetype)) + 
  scale_color_manual(values=alpha(c("grey", "blue", "green"),c(0.2, 0.5, 0.5))) + ylab("Ks between X and Y genes") + xlab("gene start position on Y chromosome") +
  theme_classic() +
  geom_vline(xintercept = 45e6, linetype = "dashed")
ggsave("Ks_labels_ycoord.pdf",h=5,w=7)

```

genome Ks plot for the hell of it
```{r}
ks_mb<-ks_data_separated %>% mutate_at(c('start1', 'start2'), as.numeric) %>% 
  mutate(start1_mb = (start1/1e6)) %>%
  mutate(start2_mb = (start2/1e6))
library(slider)
windowmaker_1<- function(df,win,ks){df %>% filter(Ks<ks) %>%
  group_by(chr1) %>%
  arrange(name1,chr1, start1_mb) %>%
  mutate(Ks_win = slider::slide_dbl(Ks, median, .before = win/2, .after = win/2,.step = 1,.complete =T))}

windowmaker_2<- function(df,win,ks){df %>% filter(Ks<ks) %>%
  group_by(chr2) %>%
  arrange(name2,chr2, start2_mb) %>%
  mutate(Ks_win = slider::slide_dbl(Ks, median, .before = win/2, .after = win/2,.step = 1,.complete =T))}

ks_win100<-windowmaker_1(ks_mb,win = 100,0.3)
ks_win100_x<-windowmaker_2(ks_mb, win = 100, 0.3)

ks_win50<-windowmaker_1(ks_mb,win = 50,0.3)
ks_win20_x<-windowmaker_2(ks_mb, win = 20, 0.3)
library(RColorBrewer)
ggplot(filter(ks_win50,!grepl("scaffold", chr1)), aes(start1_mb,group = chr1)) +  
   geom_line(aes(y=Ks_win, colour = chr1)) +
   facet_grid(~chr1,scales="free_x") +
   ylim(0,0.05) +
  scale_color_brewer(palette = "Dark2") +
  xlab("Y-bearing Haplotype\nposition(Mb)") +
  ylab("Median Ks\n(100 gene windows)") 

ggsave("Ks_coords_allchr_TX_hap2_20win.jpg", h = 5, w = 9)


ggplot(filter(ks_win20_x,!grepl("scaffold", chr2)), aes(start2_mb,group = chr2)) +  
   geom_line(aes(y=Ks_win, colour = chr2)) +
   facet_grid(~chr2,scales="free_x") +
   ylim(0,0.05) +
  scale_color_brewer(palette = "Dark2") +
  xlab("X-bearing Haplotype\nposition(Mb)") +
  ylab("Median Ks\n(100 gene windows)") 
ggsave("Ks_coords_allchr_TX_hap1_20win.pdf", h = 5, w = 9)

```

```{r}
#Ks of DNA badgenes ?!

bad_ks<- filter(ks_data_separated, id_tx_mat%in%badgenes_dna$id_tx_mat & id_tx_pat%in%badgenes_dna$id_tx_pat) %>% filter(chr1  =="Y" & chr2 =="X")
### recall, Ks calculated b/w genes - much of these DNA reads are not genic!!!
bad_ks_rna<- filter(ks_data_separated, id_tx_mat%in%badgenes_x$id_tx_mat & id_tx_pat%in%badgenes$Geneid) %>% filter(chr1  =="Y" & chr2 =="X")

baddf<-setdiff(badgenes_x$id_tx_mat,badgenes_dna$id_tx_mat) # 960 genes in RNA are not mismapping in the DNA data - maternal
reversebaddf<-setdiff(badgenes_dna$id_tx_mat, badgenes_x$id_tx_mat) #404 genes in dna are not mismapping in the RNA

baddf2<-setdiff(badgenes$Geneid,badgenes_dna$id_tx_pat) #1144
reversebaddf2<-setdiff(badgenes_dna$id_tx_pat,badgenes$Geneid) #405
```

